diff -ur srcold/apinger.c src/apinger.c
--- srcold/apinger.c	2013-08-05 19:34:15.000000000 +0000
+++ src/apinger.c	2013-08-05 22:24:27.000000000 +0000
@@ -28,6 +28,9 @@
 #ifdef HAVE_UNISTD_H
 # include <unistd.h>
 #endif
+#ifdef HAVE_SIGNAL_H
+# include <signal.h>
+#endif
 #ifdef HAVE_STRING_H
 # include <string.h>
 #endif
@@ -938,6 +955,7 @@
 			reload_request=0;
 			logit("SIGHUP received, reloading configuration.");
 			reload_config();
+			signal(SIGHUP, signal_handler);
 		}
 		for(t=targets;t;t=t->next){
 			for(aal=t->active_alarms;aal;aal=aal->next){
@@ -970,6 +988,7 @@
 				logit("SIGUSR1 received, writting status.");
 				write_status();
 			}
+			signal(SIGUSR1, signal_handler);
 		}
 		if (config->rrd_interval){
 			if (scheduled_event(&next_rrd_update,&cur_time,config->rrd_interval)){
diff -ur srcold/apinger.h src/apinger.h
--- srcold/apinger.h	2013-08-05 19:34:15.000000000 +0000
+++ src/apinger.h	2013-08-05 19:37:20.000000000 +0000
@@ -143,6 +143,7 @@
 
 const char * subst_macros(const char *string,struct target *t,struct alarm_cfg *a,int on);
 
+void signal_handler(int);
 extern volatile int interrupted_by;
 extern volatile int reload_request;
 extern volatile int status_request;
diff -ur srcold/main.c src/main.c
--- srcold/main.c	2013-08-05 19:34:15.000000000 +0000
+++ src/main.c	2013-08-05 19:37:32.000000000 +0000
@@ -110,15 +110,15 @@
 void signal_handler(int signum){
 
 	if (signum==SIGPIPE){
-		signal(SIGPIPE,signal_handler);
+		signal(SIGPIPE, SIG_IGN);
 		sigpipe_received=1;
 	}
 	else if (signum==SIGHUP){
-		signal(SIGHUP,signal_handler);
+		signal(SIGHUP, SIG_IGN);
 		reload_request=1;
 	}
 	else if (signum==SIGUSR1){
-		signal(SIGUSR1,signal_handler);
+		signal(SIGUSR1, SIG_IGN);
 		status_request=1;
 	}
 	else{
diff -ur srcold/rrd.c src/rrd.c
--- srcold/rrd.c	2013-08-05 19:34:15.000000000 +0000
+++ src/rrd.c	2013-08-05 19:45:03.000000000 +0000
@@ -30,6 +30,9 @@
 #ifdef HAVE_UNISTD_H
 # include <unistd.h>
 #endif
+#ifdef HAVE_SIGNAL_H
+# include <signal.h>
+#endif
 #ifdef HAVE_STRING_H
 # include <string.h>
 #endif
@@ -136,6 +139,7 @@
 		sigpipe_received=0;
 		if (rrdtool_pipe) pclose(rrdtool_pipe);
 		rrdtool_pipe=NULL;
+		signal(SIGPIPE, signal_handler);
 	}
 	for(t=targets;t!=NULL;t=t->next){
 		if (t->config->rrd_filename==NULL) continue;
