--- srcold/apinger.c	2010-12-23 18:44:00.000000000 +0000
+++ src/apinger.c	2010-12-23 18:45:06.000000000 +0000
@@ -805,7 +805,6 @@
 time_t tm;
 int i,qp,really_lost;
 char *buf1,*buf2;
-int err=0;
 
 	if (config->status_file==NULL) return;
 	
@@ -867,7 +866,7 @@
 		if (t->recently_lost!=really_lost){
 			logit("Target \"%s\": Lost packet count mismatch (%i(recently_lost) != %i(really_lost))!",t->name,t->recently_lost,really_lost);
 			logit("Target \"%s\": Received packets buffer: %s %s\n",t->name,buf2,buf1);
-			err=1;
+			//t->recently_lost = really_lost = 0;
 		}
 		free(buf1);
 		free(buf2);
@@ -875,7 +874,6 @@
 		fprintf(f,"\n");
 	}
 	fclose(f);
-	if (err) abort();
 }
 
 void main_loop(void){
