diff -ur ../radvd-1.9.1.old/device-bsd44.c ./device-bsd44.c
--- ../radvd-1.9.1.old/device-bsd44.c	2013-03-20 12:06:55.000000000 +0000
+++ ./device-bsd44.c	2013-03-20 12:11:26.000000000 +0000
@@ -17,8 +17,21 @@
 #include "radvd.h"
 #include "defaults.h"
 #include "pathnames.h"		/* for PATH_PROC_NET_IF_INET6 */
+#include "rtsock.h"
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <net/route.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+
+#include <netinet/ip_carp.h>
 
 static uint8_t ll_prefix[] = { 0xfe, 0x80 };
+extern struct Interface *IfaceList;
 
 /*
  * this function gets the hardware type and address of an interface,
@@ -136,6 +149,130 @@
 	return -1;
 }
 
+int
+setup_carpinfo(struct Interface *iface)
+{
+	struct ifaddrs *addresses = 0, *ifa;
+	struct AdvPrefix *prefix, **pptr;
+	struct Interface *parent;
+	struct ifreq	ifr;
+	struct carpreq carpr;
+	struct in6_addr nw;
+	unsigned int pix = 0;
+
+	if (iface->AdvPrefixList == NULL ||
+	    (iface->AdvPrefixList != NULL &&
+	    iface->AdvPrefixList->next != NULL)) {
+		flog(LOG_ERR, "%s must have one prefix configured", iface->Name);
+		return -1;
+	}
+
+	strncpy(ifr.ifr_name, iface->Name, IFNAMSIZ-1);
+	ifr.ifr_name[IFNAMSIZ-1] = '\0';
+
+	memset((char *)&carpr, 0, sizeof(struct carpreq));
+	ifr.ifr_data = (caddr_t)&carpr;
+	if (ioctl(sock, SIOCGVH, &ifr) < 0) {
+		flog(LOG_ERR, "ioctl SIOCGVH failed: %s(%d)", strerror(errno), errno);
+		return -1;
+	}
+
+	if (getifaddrs(&addresses) != 0)
+	{
+		flog(LOG_ERR, "getifaddrs failed: %s(%d)", strerror(errno), errno);
+		return -1;
+	}
+
+	parent = NULL;
+	for (ifa = addresses; ifa != NULL; ifa = ifa->ifa_next)
+	{
+		struct sockaddr_in6 *s6 = (struct sockaddr_in6 *)ifa->ifa_addr;
+		struct sockaddr_in6 *mask = (struct sockaddr_in6 *)ifa->ifa_netmask;
+
+		if (strlen(ifa->ifa_name) != strlen(iface->Name))
+			continue;
+
+		if (strcmp(ifa->ifa_name, iface->Name) == 0)
+			continue;	/* looking for !carp */
+
+		if (ifa->ifa_addr == NULL)
+			continue;
+
+		if (ifa->ifa_addr->sa_family != AF_INET6)
+			continue;
+
+		if (IN6_IS_ADDR_LINKLOCAL(&s6->sin6_addr))
+			continue;
+
+		nw = get_prefix6(&s6->sin6_addr, &mask->sin6_addr);
+		prefix = iface->AdvPrefixList;
+		while (prefix)
+		{
+			if (addr_match(&prefix->Prefix, &nw,
+				    prefix->PrefixLen)) {
+				flog(LOG_ERR, "Carp prefix found on %s", ifa->ifa_name);
+				break;
+			}
+			prefix = prefix->next;
+		}
+		if (prefix == NULL)
+			continue;
+		pix = if_nametoindex(ifa->ifa_name);
+
+		flog(LOG_ERR, "finding interface %s", ifa->ifa_name);
+		/* find the interface this prefix is on */
+		parent = IfaceList;
+		while (parent)
+		{
+			if (!strcmp(parent->Name, ifa->ifa_name))
+				break;	/* found it */
+			parent = parent->next;
+		}
+		break;
+	}
+	freeifaddrs(addresses);
+
+	if (parent != NULL) {
+		if (parent->if_hwaddr_len != 48) {
+			flog(LOG_ERR, "if_hwaddr_len %d, expected 48", parent->if_hwaddr_len);
+			return -1;
+		}
+
+		iface->if_maxmtu = parent->if_maxmtu;
+		iface->if_prefix_len = parent->if_prefix_len;
+
+		/* remove the prefix from the carp parent, if it exists */
+		pptr = &parent->AdvPrefixList;
+		while (*pptr) {
+			prefix = *pptr;
+
+			if (IN6_ARE_ADDR_EQUAL(&prefix->Prefix, &iface->AdvPrefixList->Prefix)) {
+				*pptr = prefix->next;
+				free(prefix);
+				break;
+			}
+			pptr = &prefix->next;
+		}
+	}
+
+	iface->carp_p_index = pix;
+	flog(LOG_ERR, "PIX = %d", pix);
+	iface->if_index = if_nametoindex(iface->Name);
+	/* Only ethernet supported at the moment */
+	iface->if_hwaddr_len = 48;
+	iface->if_hwaddr[0] = 0x00;
+	iface->if_hwaddr[1] = 0x00;
+	iface->if_hwaddr[2] = 0x5e;
+	iface->if_hwaddr[3] = 0x00;
+	iface->if_hwaddr[4] = 0x01;
+	iface->if_hwaddr[5] = carpr.carpr_vhid;
+
+	if (carpr.carpr_state != 2 /* master */)
+		iface->LinkDown = 1;
+
+	return 0;
+}
+
 /*
  * Saves the first link local address seen on the specified interface to iface->if_addr
  *
@@ -152,6 +289,9 @@
 
 	for (ifa = addresses; ifa != NULL; ifa = ifa->ifa_next)
 	{
+		if (strlen(ifa->ifa_name) != strlen(iface->Name))
+			continue;
+
 		if (strcmp(ifa->ifa_name, iface->Name) != 0)
 			continue;
 
@@ -225,6 +365,85 @@
 	return -1;
 }
 
+int carp_master(const char *iface)
+{
+	struct ifreq	ifr;
+	struct carpreq carpr;
+
+	strncpy(ifr.ifr_name, iface, IFNAMSIZ-1);
+	ifr.ifr_name[IFNAMSIZ-1] = '\0';
+
+	memset((char *)&carpr, 0, sizeof(struct carpreq));
+	ifr.ifr_data = (caddr_t)&carpr;
+	if (ioctl(sock, SIOCGVH, &ifr) < 0)
+		return -1;
+
+	return carpr.carpr_state == 2;
+}
+
+void process_rtsock_msg(int sock)
+{
+	char buf[2048];
+	struct Interface *iface;
+	struct if_msghdr *ifm;
+	struct rt_msghdr *rtm;
+	int n;
+
+	n = read(sock, buf, sizeof(buf));
+	if (n < 0) {
+		if (errno != EINTR && errno != EAGAIN)
+			perror("read(PF_ROUTE)");
+		return;
+	}
+
+	rtm = (struct rt_msghdr *) buf;
+	if (rtm->rtm_version != RTM_VERSION) {
+		dlog(LOG_DEBUG, 3, "Routing message version %d not "
+			"understood\n", rtm->rtm_version);
+		return;
+	}
+	switch (rtm->rtm_type) {
+	case RTM_IFINFO:
+		ifm = (struct if_msghdr *) rtm;
+		dlog(LOG_DEBUG, 3, "Routing message version %d iface %d\n",
+			rtm->rtm_version, ifm->ifm_index);
+		iface = IfaceList;
+		while (iface) {
+			if (ifm->ifm_index == iface->if_index)
+				break;
+			iface = iface->next;
+		}
+		if (iface == NULL)
+			break;	/* not ours */
+
+		switch (ifm->ifm_data.ifi_link_state) {
+			case LINK_STATE_DOWN:
+				iface->LinkDown = 1;
+				break;
+			case LINK_STATE_UP:
+				iface->LinkDown = 0;
+				break;
+			default:
+				/* dont know... assume up */
+				iface->LinkDown = 1;
+				break;
+		}
+		dlog(LOG_DEBUG, 3, "Interface %s changed to %s",
+		    iface->Name, iface->LinkDown ? "DOWN":"UP");
+		/* Should check other flags like IFF_UP */
+		break;
+	}
+}
+
+int rt_socket(void)
+{
+	int sock;
+
+	sock = socket(PF_ROUTE, SOCK_RAW, 0);
+
+	return sock;
+}
+
 int
 set_interface_curhlim(const char *iface, uint8_t hlim)
 {
diff -ur ../radvd-1.9.1.old/device-common.c ./device-common.c
--- ../radvd-1.9.1.old/device-common.c	2013-03-20 12:06:55.000000000 +0000
+++ ./device-common.c	2013-03-20 12:06:59.000000000 +0000
@@ -46,14 +46,14 @@
 		return (-1);
 	}
 
-	if (! iface->UnicastOnly && !(ifr.ifr_flags & IFF_MULTICAST))
+	if (! iface->IsCarp && ! iface->UnicastOnly && !(ifr.ifr_flags & IFF_MULTICAST))
 	{
 		flog(LOG_WARNING, "interface %s does not support multicast",
 			iface->Name);
 		flog(LOG_WARNING, "   do you need to add the UnicastOnly flag?");
 	}
 
-	if (! iface->UnicastOnly && !(ifr.ifr_flags & IFF_BROADCAST))
+	if (! iface->IsCarp && ! iface->UnicastOnly && !(ifr.ifr_flags & IFF_BROADCAST))
 	{
 		flog(LOG_WARNING, "interface %s does not support broadcast",
 			iface->Name);
diff -ur ../radvd-1.9.1.old/gram.y ./gram.y
--- ../radvd-1.9.1.old/gram.y	2013-03-20 12:06:55.000000000 +0000
+++ ./gram.y	2013-03-20 12:06:59.000000000 +0000
@@ -33,7 +33,6 @@
 static void yyerror(char *msg);
 static int countbits(int b);
 static int count_mask(struct sockaddr_in6 *m);
-static struct in6_addr get_prefix6(struct in6_addr const *addr, struct in6_addr const *mask);
 
 #if 0 /* no longer necessary? */
 #ifndef HAVE_IN6_ADDR_S6_ADDR
@@ -168,6 +167,10 @@
 				iface2 = iface2->next;
 			}
 
+#ifdef __FreeBSD__
+			if (carp_master(iface->Name) != -1)
+				iface->IsCarp = 1;
+#endif
 			if (check_device(iface) < 0) {
 				if (iface->IgnoreIfMissing) {
 					dlog(LOG_DEBUG, 4, "interface %s did not exist, ignoring the interface", iface->Name);
@@ -180,6 +183,8 @@
 			if (setup_deviceinfo(iface) < 0)
 				if (!iface->IgnoreIfMissing)
 				ABORT;
+			if (iface->IsCarp && setup_carpinfo(iface) < 0)
+				ABORT;
 			if (check_iface(iface) < 0)
 				if (!iface->IgnoreIfMissing)
 				ABORT;
@@ -191,6 +196,7 @@
 				ABORT;
 
 			dlog(LOG_DEBUG, 4, "interface definition for %s is ok", iface->Name);
+			dlog(LOG_DEBUG, 4, "%s: if_index %d", iface->Name, iface->if_index);
 
 			iface->next = IfaceList;
 			IfaceList = iface;
@@ -940,7 +946,6 @@
 	return count;
 }
 
-static
 struct in6_addr get_prefix6(struct in6_addr const *addr, struct in6_addr const *mask)
 {
 	struct in6_addr prefix = *addr;
diff -ur ../radvd-1.9.1.old/process.c ./process.c
--- ../radvd-1.9.1.old/process.c	2013-03-20 12:06:55.000000000 +0000
+++ ./process.c	2013-03-20 12:07:31.000000000 +0000
@@ -21,14 +21,12 @@
 		       int len, struct sockaddr_in6 *);
 static void process_ra(struct Interface *, unsigned char *msg, int len,
 	struct sockaddr_in6 *);
-static int  addr_match(struct in6_addr *a1, struct in6_addr *a2,
-	int prefixlen);
 
 void
 process(struct Interface *ifacel, unsigned char *msg, int len,
 	struct sockaddr_in6 *addr, struct in6_pktinfo *pkt_info, int hoplimit)
 {
-	struct Interface *iface;
+	struct Interface *iface, *carp = NULL;
 	struct icmp6_hdr *icmph;
 	char addr_str[INET6_ADDRSTRLEN];
 
@@ -104,8 +102,14 @@
 		{
 			break;
 		}
+		if (iface->carp_p_index == pkt_info->ipi6_ifindex)
+		{
+			carp = iface;
+		}
 	}
 
+	if (iface == NULL && carp != NULL)
+		iface = carp;
 	if (iface == NULL)
 	{
 		dlog(LOG_DEBUG, 2, "received packet from unknown interface: %d",
@@ -473,7 +477,7 @@
 	}
 }
 
-static int
+int
 addr_match(struct in6_addr *a1, struct in6_addr *a2, int prefixlen)
 {
 	unsigned int pdw;
diff -ur ../radvd-1.9.1.old/radvd.c ./radvd.c
--- ../radvd-1.9.1.old/radvd.c	2013-03-20 12:06:55.000000000 +0000
+++ ./radvd.c	2013-03-20 12:06:59.000000000 +0000
@@ -17,6 +17,7 @@
 #include "includes.h"
 #include "radvd.h"
 #include "pathnames.h"
+#include "rtsock.h"
 
 #ifdef HAVE_NETLINK
 #include "netlink.h"
@@ -374,8 +375,8 @@
 	fds[1].events = POLLIN;
 	fds[1].revents = 0;
 #else
-	fds[1].fd = -1;
-	fds[1].events = 0;
+	fds[1].fd = rt_socket();
+	fds[1].events = POLLIN;
 	fds[1].revents = 0;
 #endif
 
@@ -418,14 +419,17 @@
 						&rcv_addr, pkt_info, hoplimit);
 				}
 			}
-#ifdef HAVE_NETLINK
 			if (fds[1].revents & (POLLERR | POLLHUP | POLLNVAL)) {
 				flog(LOG_WARNING, "socket error on fds[1].fd");
 			}
 			else if (fds[1].revents & POLLIN) {
+#ifdef HAVE_NETLINK
 				process_netlink_msg(fds[1].fd);
-			}
+#else
+				process_rtsock_msg(fds[1].fd);
+
 #endif
+			}
 		}
 		else if ( rc == 0 ) {
 			if (next)
diff -ur ../radvd-1.9.1.old/radvd.h ./radvd.h
--- ../radvd-1.9.1.old/radvd.h	2013-03-20 12:06:55.000000000 +0000
+++ ./radvd.h	2013-03-20 12:06:59.000000000 +0000
@@ -66,6 +66,10 @@
 	int			AdvSourceLLAddress;
 	int			UnicastOnly;
 
+	/* FreeBSD Carp */
+	int			IsCarp;
+	unsigned int		carp_p_index;
+
 	/* Mobile IPv6 extensions */
 	int			AdvIntervalOpt;
 	int			AdvHomeAgentInfo;
@@ -86,6 +90,7 @@
 
 	/* Info whether this interface has failed in the past (and may need to be reinitialized) */
 	int			HasFailed;
+	int			LinkDown;
 
 	struct Interface	*next;
 };
@@ -179,6 +184,7 @@
 
 /* gram.y */
 int yyparse(void);
+struct in6_addr get_prefix6(struct in6_addr const *, struct in6_addr const *);
 
 /* scanner.l */
 int yylex(void);
@@ -196,6 +202,7 @@
 
 /* device.c */
 int setup_deviceinfo(struct Interface *);
+int setup_carpinfo(struct Interface *);
 int check_device(struct Interface *);
 int setup_linklocal_addr(struct Interface *);
 int setup_allrouters_membership(struct Interface *);
@@ -206,6 +213,7 @@
 int set_interface_curhlim(const char *, uint8_t);
 int set_interface_reachtime(const char *, uint32_t);
 int set_interface_retranstimer(const char *, uint32_t);
+int carp_master(const char *);
 
 /* interface.c */
 void iface_init_defaults(struct Interface *);
@@ -225,6 +233,8 @@
 /* process.c */
 void process(struct Interface *, unsigned char *, int,
 	struct sockaddr_in6 *, struct in6_pktinfo *, int);
+int  addr_match(struct in6_addr *a1, struct in6_addr *a2,
+	int prefixlen);
 
 /* recv.c */
 int recv_rs_ra(unsigned char *, struct sockaddr_in6 *, struct in6_pktinfo **, int *);
diff -ur ../radvd-1.9.1.old/rtsock.h ./rtsock.h
--- /dev/null
+++ rtsock.h
@@ -0,0 +1,20 @@
+/*
+ *
+ *   Authors:
+ *    Lars Fenneberg           <lf@elemental.net>
+ *    Reuben Hawkins           <reubenhwk@gmail.com>
+ *
+ *   This software is Copyright 1996,1997 by the above mentioned author(s),
+ *   All Rights Reserved.
+ *
+ *   The license which is distributed with this software in the file COPYRIGHT
+ *   applies to this software. If your distribution is missing this file, you
+ *   may request it from <pekkas@netcore.fi>.
+ *
+ */
+
+#pragma once
+
+void process_rtsock_msg(int sock);
+int rt_socket(void);
+
diff -ur ../radvd-1.9.1.old/send.c ./send.c
--- ../radvd-1.9.1.old/send.c	2013-03-20 12:06:55.000000000 +0000
+++ ./send.c	2013-03-20 12:06:59.000000000 +0000
@@ -134,6 +134,13 @@
 	size_t len = 0;
 	ssize_t err;
 
+#ifdef __FreeBSD__
+	if (iface->IsCarp && carp_master(iface->Name) != 1) {
+		dlog(LOG_DEBUG, 4, "interface %s is not carp master", iface->Name);
+		/* not really a 'success', but we need to schedule new timers.. */
+		return 0;
+	} else
+#endif
 	/* First we need to check that the interface hasn't been removed or deactivated */
 	if(check_device(iface) < 0) {
 		if (iface->IgnoreIfMissing)  /* a bit more quiet warning message.. */
