--- send.c.orig	2017-11-08 18:38:09.371714000 -0600
+++ send.c	2017-11-08 18:53:22.896923000 -0600
@@ -832,9 +832,11 @@
 	iov.iov_len = sb->used;
 	iov.iov_base = (caddr_t)sb->buffer;
 
-	char __attribute__((aligned(8))) chdr[CMSG_SPACE(sizeof(struct in6_pktinfo))];
-	memset(chdr, 0, sizeof(chdr));
-	struct cmsghdr *cmsg = (struct cmsghdr *)chdr;
+	int cmsglen = CMSG_SPACE(sizeof(struct in6_pktinfo));
+	struct cmsghdr *cmsg = (struct cmsghdr *)malloc(cmsglen);
+	if (cmsg == NULL)
+		return (-1);
+	memset(cmsg, 0, cmsglen);
 
 	cmsg->cmsg_len = CMSG_LEN(sizeof(struct in6_pktinfo));
 	cmsg->cmsg_level = IPPROTO_IPV6;
@@ -856,9 +858,12 @@
 	mhdr.msg_iov = &iov;
 	mhdr.msg_iovlen = 1;
 	mhdr.msg_control = (void *)cmsg;
-	mhdr.msg_controllen = sizeof(chdr);
+	mhdr.msg_controllen = cmsglen;
 
-	return sendmsg(sock, &mhdr, 0);
+	int err = sendmsg(sock, &mhdr, 0);
+	free(cmsg);
+
+	return (err);
 }
 
 static int schedule_option_prefix(struct in6_addr const *dest, struct Interface const *iface, struct AdvPrefix const *prefix)
