commit 073a6baceffc4a363eac9369cc036fc7b19b919e
Author: jim-p <jimp@netgate.com>
Date:   Tue May 30 09:51:01 2023 -0400

    Fix mac_procs incorrect references. Fixes #14425

diff --git a/src/usr/local/pfSense/include/www/system_advanced_admin.inc b/src/usr/local/pfSense/include/www/system_advanced_admin.inc
index 3ecb52f528..d3c9f67bcb 100644
--- a/src/usr/local/pfSense/include/www/system_advanced_admin.inc
+++ b/src/usr/local/pfSense/include/www/system_advanced_admin.inc
@@ -163,7 +163,7 @@ function doAdvancedAdminPOST($post, $json = false) {
 		}
 
 		if ($post['max_procs']) {
-			config_set_path('system/webgui/max_proc', $post['max_proc']);
+			config_set_path('system/webgui/max_procs', $post['max_procs']);
 			$restart_webgui = true;
 		}
 
