diff --git a/src/etc/inc/filter.inc b/src/etc/inc/filter.inc
index d36d6df2e2..2265995a56 100644
--- a/src/etc/inc/filter.inc
+++ b/src/etc/inc/filter.inc
@@ -430,7 +430,6 @@ function filter_configure_sync($delete_states_if_needed = true) {
 	$rules .= filter_setup_logging_interfaces();
 	$rules .= "\n";
 	$rules .= "set skip on pfsync0\n";
-	$rules .= "set keepcounters\n";
 	$rules .= "\n";
 	update_filter_reload_status(gettext("Setting up SCRUB information"));
 	$rules .= filter_generate_scrubing();
