diff --git a/src/usr/local/www/interfaces_groups.php b/src/usr/local/www/interfaces_groups.php
index b002adedcf..62bdaeaf72 100644
--- a/src/usr/local/www/interfaces_groups.php
+++ b/src/usr/local/www/interfaces_groups.php
@@ -98,7 +98,7 @@ display_top_tabs($tab_array);
 
 		unset($iflist);
 		$memberses = implode(", ", $memberses_arr);
-		echo $memberses;
+		echo htmlspecialchars($memberses);
 		if (count($members_arr) >= 10) {
 			echo '&hellip;';
 		}
diff --git a/src/usr/local/www/interfaces_groups_edit.php b/src/usr/local/www/interfaces_groups_edit.php
index 6bcb30eec6..34b196c465 100644
--- a/src/usr/local/www/interfaces_groups_edit.php
+++ b/src/usr/local/www/interfaces_groups_edit.php
@@ -36,28 +36,38 @@ $pgtitle = array(gettext("Interfaces"), gettext("Interface Groups"), gettext("Ed
 $pglinks = array("", "interfaces_groups.php", "@self");
 $shortcut_section = "interfaces";
 
+$interface_list = get_configured_interface_with_descr(true);
+$interface_list_disabled = get_configured_interface_with_descr(true);
+/* hide VTI interfaces, see https://redmine.pfsense.org/issues/11134 */
+foreach ($interface_list as $if => $ifdescr) {
+	if (substr(get_real_interface($if), 0, 5) == "ipsec") {
+		unset($interface_list[$if]);
+	}
+}
+
 init_config_arr(array('ifgroups', 'ifgroupentry'));
 $a_ifgroups = &$config['ifgroups']['ifgroupentry'];
 $id = $_REQUEST['id'];
 
 if (isset($id) && $a_ifgroups[$id]) {
+	/* Cleanup invalid group members (Deleted interfaces, etc.)
+	 * https://redmine.pfsense.org/issues/15778 */
+	$validmembers = [];
+	foreach (explode(" ", array_get_path($a_ifgroups[$id], 'members', "")) as $ifname) {
+		if (array_key_exists($ifname, $interface_list)) {
+			$validmembers[] = $ifname;
+		}
+	}
+	array_set_path($a_ifgroups[$id], 'members', implode(" ", $validmembers));
+
 	$pconfig['ifname'] = $a_ifgroups[$id]['ifname'];
 	$pconfig['members'] = $a_ifgroups[$id]['members'];
 	$pconfig['descr'] = html_entity_decode($a_ifgroups[$id]['descr']);
 }
 
-$interface_list = get_configured_interface_with_descr(true);
-$interface_list_disabled = get_configured_interface_with_descr(true);
 $ifname_allowed_chars_text = gettext("Only letters (A-Z), digits (0-9) and '_' are allowed.");
 $ifname_no_digit_text = gettext("The group name cannot start or end with a digit.");
 
-/* hide VTI interfaces, see https://redmine.pfsense.org/issues/11134 */
-foreach ($interface_list as $if => $ifdescr) {
-	if (substr(get_real_interface($if), 0, 5) == "ipsec") {
-		unset($interface_list[$if]);
-	}
-}
-
 if ($_POST['save']) {
 	unset($input_errors);
 	$pconfig = $_POST;
@@ -114,8 +124,20 @@ if ($_POST['save']) {
 		}
 	}
 
-	if (isset($_POST['members'])) {
-		$members = implode(" ", $_POST['members']);
+	/* Ensure submitted interfaces exist in the configuration, filter
+	 * invalid entries from selected interface list.
+	 * https://redmine.pfsense.org/issues/15778 */
+	$validmembers = [];
+	foreach ($_POST['members'] as $ifname) {
+		if (array_key_exists($ifname, $interface_list)) {
+			$validmembers[] = $ifname;
+		} else {
+			$input_errors[] = gettext("Submission contained an invalid interface");
+		}
+	}
+
+	if (!empty($validmembers)) {
+		$members = implode(" ", $validmembers);
 	} else {
 		$members = "";
 	}
