diff --git a/src/etc/inc/pfsense-utils.inc b/src/etc/inc/pfsense-utils.inc
index 365174042f..3704c52eaf 100644
--- a/src/etc/inc/pfsense-utils.inc
+++ b/src/etc/inc/pfsense-utils.inc
@@ -3078,18 +3078,33 @@ function filter_rules_sort() {
 	config_set_path('filter/rule', $rules);
 }
 function filter_rules_compare($a, $b) {
-	if ((isset($a['floating']) && isset($b['floating'])) ||
-	    (isset($b['ethernet']) && isset($b['ethernet']))) {
-		return $a['seq'] - $b['seq'];
-	} else if (isset($a['floating']) || isset($a['ethernet'])) {
-		return -1;
-	} else if (isset($b['floating']) || isset($b['ethernet'])) {
+	if (isset($a['ethernet'])) {
+		// Ethernet rules always come first.
+		if (!isset($b['ethernet'])) {
+			return -1;
+		}
+	} elseif (isset($a['floating'])) {
+		// Floating rules are placed after Ethernet rules and before Interface rules.
+		if (isset($b['ethernet'])) {
+			return 1;
+		} elseif (!isset($b['floating'])) {
+			return -1;
+		}
+	} elseif (isset($b['ethernet']) || isset($b['floating'])) {
+		// Interface rules are placed after Ethernet and Floating rules.
 		return 1;
-	} else if ($a['interface'] == $b['interface']) {
-		return $a['seq'] - $b['seq'];
-	} else {
+	} elseif ($a['interface'] != $b['interface']) {
+		// Interface rules for the same interface are grouped together.
 		return compare_interface_friendly_names($a['interface'], $b['interface']);
 	}
+
+	// Use the sequence number as a last resort.
+	if (isset($a['seq']) && isset($b['seq'])) {
+		return ($a['seq'] <=> $b['seq']);
+	}
+
+	// Nothing to compare.
+	return 0;
 }
 
 function generate_ipv6_from_mac($mac) {
