commit 1d33b166a44318f893766f55b8d092f70be6380b
Author: jim-p <jimp@netgate.com>
Date:   Fri Dec 6 12:00:33 2024 -0500

    Close PHP session when exiting early. Fixes #15873

diff --git a/src/etc/inc/authgui.inc b/src/etc/inc/authgui.inc
index b3a9c5e429..3ea6c449c3 100644
--- a/src/etc/inc/authgui.inc
+++ b/src/etc/inc/authgui.inc
@@ -71,12 +71,14 @@ if (!isAllowedPage($_SERVER['REQUEST_URI'])) {
 
 		$username = get_config_user();
 		log_error("{$username} attempted to access {$_SERVER['SCRIPT_NAME']} but does not have access to that page. Redirecting to {$page}.");
+		phpsession_end();
 
 		exit;
 	} else {
 		// add this so they don't get stuck on the logout page when they have no permissions.
 		$_SESSION["Logged_In"] = false;
 		display_error_form("201", gettext("No page assigned to this user! Click here to logout."));
+		phpsession_end();
 
 		exit;
 	}
@@ -92,6 +94,8 @@ if (!$_SESSION['Post_Login']) {
 	$_SESSION['Post_Login'] = true;
 	require_once("functions.inc");
 	pfSenseHeader($_SERVER['REQUEST_URI']);
+	phpsession_end();
+
 	exit;
 }
 
