diff --git a/src/etc/inc/pfsense-utils.inc b/src/etc/inc/pfsense-utils.inc
index eada3dc909..03acb30cbc 100644
--- a/src/etc/inc/pfsense-utils.inc
+++ b/src/etc/inc/pfsense-utils.inc
@@ -2599,7 +2599,7 @@ function pfs_version_compare($cur_time, $cur_text, $remote) {
 function process_alias_urltable($name, $type, $url, $freq, $forceupdate=false, $validateonly=false) {
 	global $g, $config;
 
-	if (!is_validaliasname($name) || !isURL($url)) {
+	if (!is_validaliasname($name) || !filter_var($url, FILTER_VALIDATE_URL)) {
 		return false;
 	}
 
