commit 247909ff5387200fb31c76f13e7702a8fbdc06f4
Author: jim-p <jimp@netgate.com>
Date:   Wed Apr 24 15:43:29 2024 -0400

    Fix resolv.conf logic for DNS Forwarder. Fixes #15434

diff --git a/src/etc/inc/system.inc b/src/etc/inc/system.inc
index 567b3ff0aa..d087183fdd 100644
--- a/src/etc/inc/system.inc
+++ b/src/etc/inc/system.inc
@@ -1713,13 +1713,13 @@ function get_dns_nameservers($add_v6_brackets = false, $hostns=true) {
 		echo "get_dns_nameservers() being called $mt\n";
 	}
 
-	if ((((config_path_enabled('dnsmasq')) &&
+	if ((((config_path_enabled('dnsmasq') &&
 		  (config_get_path('dnsmasq/port', '53') == '53') &&
 		  in_array("lo0", explode(",", config_get_path('dnsmasq/interface', 'lo0'))))) ||
 	    (config_path_enabled('unbound') &&
 		 (config_get_path('unbound/port', '53') == '53') &&
 		 (in_array("lo0", explode(",", config_get_path('unbound/active_interface', 'lo0'))) ||
-		  in_array("all", explode(",", config_get_path('unbound/active_interface', 'all')), true))) &&
+		  in_array("all", explode(",", config_get_path('unbound/active_interface', 'all')), true)))) &&
 	    (config_get_path('system/dnslocalhost') != 'remote')) {
 		$dns_nameservers[] = "127.0.0.1";
 		$dns_nameservers[] = $add_v6_brackets ? '[::1]' : '::1';
