diff --git a/src/etc/inc/upgrade_config.inc b/src/etc/inc/upgrade_config.inc
index 8113977082..80ac91cec2 100644
--- a/src/etc/inc/upgrade_config.inc
+++ b/src/etc/inc/upgrade_config.inc
@@ -6714,6 +6714,7 @@ function upgrade_227_to_228() {
 					if (count($phase1['encryption']['item']) == 0) {
 						/* Only disable P1 if there are no valid encryption options left. */
 						$phase1['disabled'] = true;
+						unset($phase1['encryption']);
 						file_notice("IPsec", sprintf(gettext("IPsec Phase 1 '%s' disabled after removing deprecated encryption and hashing algorithms as it has no remaining valid entries."), $phase1['descr']));
 					} else {
 						/* Let the user know that the P1 was adjusted */
diff --git a/src/usr/local/www/vpn_ipsec_phase1.php b/src/usr/local/www/vpn_ipsec_phase1.php
index 6684219162..c2e341f855 100644
--- a/src/usr/local/www/vpn_ipsec_phase1.php
+++ b/src/usr/local/www/vpn_ipsec_phase1.php
@@ -183,13 +183,14 @@ if ($p1) {
 	}
 }
 // default value for new P1 and failsafe to always have at least 1 encryption item for the Form_ListItem
-if (!is_array($pconfig['encryption']['item']) || count($pconfig['encryption']['item']) == 0) {
+
+if (count(array_get_path($pconfig, 'encryption/item', [])) == 0) {
 	$item = array();
 	$item['encryption-algorithm'] = array('name' => "aes", 'keylen' => 128);
 	$item['hash-algorithm'] = "sha256";
 	$item['prf-algorithm'] = "sha256";
 	$item['dhgroup'] = "14";
-	$pconfig['encryption']['item'][] = $item;
+	array_set_path($pconfig, 'encryption/item', [$item]);
 }
 
 if (isset($_REQUEST['dup']) && is_numericint($_REQUEST['dup'])) {
@@ -200,17 +201,20 @@ if ($_POST['save']) {
 	unset($input_errors);
 	$pconfig = $_POST;
 
+	$eitems = array_get_path($pconfig, 'encryption/item', []);
 	for($i = 0; $i < 100; $i++) {
 		if (isset($_POST['ealgo_algo'.$i])) {
-			$item = array();
-			$item['encryption-algorithm']['name'] = $_POST['ealgo_algo'.$i];
-			$item['encryption-algorithm']['keylen'] = $_POST['ealgo_keylen'.$i];
-			$item['hash-algorithm'] = $_POST['halgo'.$i];
-			$item['prf-algorithm'] = $_POST['prfalgo'.$i];
-			$item['dhgroup'] = $_POST['dhgroup'.$i];
-			$pconfig['encryption']['item'][] = $item;
+			$item = [];
+			array_set_path($item, 'encryption-algorithm', []);
+			array_set_path($item, 'encryption-algorithm/name', $_POST['ealgo_algo'.$i]);
+			array_set_path($item, 'encryption-algorithm/keylen', $_POST['ealgo_keylen'.$i]);
+			array_set_path($item, 'hash-algorithm', $_POST['halgo'.$i]);
+			array_set_path($item, 'prf-algorithm', $_POST['prfalgo'.$i]);
+			array_set_path($item, 'dhgroup', $_POST['dhgroup'.$i]);
+			$eitems[] = $item;
 		}
 	}
+	array_set_path($pconfig, 'encryption/item', $eitems);
 
 	/* input validation */
 
@@ -484,8 +488,8 @@ if ($_POST['save']) {
 		$input_errors[] = gettext("Valid arguments for IKE type are v1, v2 or auto");
 	}
 
-	foreach($pconfig['encryption']['item'] as $p1algo) {
-		if (preg_match("/aes\d+gcm/", $p1algo['encryption-algorithm']['name']) && $_POST['iketype'] != "ikev2") {
+	foreach(array_get_path($pconfig, 'encryption/item', []) as $p1algo) {
+		if (preg_match("/aes\d+gcm/", array_get_path($p1algo, 'encryption-algorithm/name', '')) && $_POST['iketype'] != "ikev2") {
 			$input_errors[] = gettext("Encryption Algorithm AES-GCM can only be used with IKEv2");
 		}
 	}
@@ -938,9 +942,10 @@ $section->addInput(new Form_Select(
 
 $form->add($section);
 
-$rowcount = count($pconfig['encryption']['item']);
+$eitems = array_get_path($pconfig, 'encryption/item', []);
+$rowcount = count($eitems);
 $section = new Form_Section('Phase 1 Proposal (Encryption Algorithm)');
-foreach($pconfig['encryption']['item'] as $key => $p1enc) {
+foreach($eitems as $key => $p1enc) {
 	$lastrow = ($counter == $rowcount - 1);
 	$group = new Form_Group($counter == 0 ? '*Encryption Algorithm' : '');
 	$group->addClass("repeatable");
@@ -948,28 +953,28 @@ foreach($pconfig['encryption']['item'] as $key => $p1enc) {
 	$group->add(new Form_Select(
 		'ealgo_algo'.$key,
 		null,
-		$p1enc['encryption-algorithm']['name'],
+		array_get_path($p1enc, 'encryption-algorithm/name', []),
 		build_eal_list()
 	))->setHelp($lastrow ? 'Algorithm' : '')->setWidth(2);
 
 	$group->add(new Form_Select(
 		'ealgo_keylen'.$key,
 		null,
-		$p1enc['encryption-algorithm']['keylen'],
+		array_get_path($p1enc, 'encryption-algorithm/keylen', []),
 		array()
 	))->setHelp($lastrow ? 'Key length' : '')->setWidth(2);
 
 	$group->add(new Form_Select(
 		'halgo'.$key,
 		'*Hash Algorithm',
-		$p1enc['hash-algorithm'],
+		array_get_path($p1enc, 'hash-algorithm'),
 		$p1_halgos
 	))->setHelp($lastrow ? 'Hash' : '')->setWidth(2);
 
 	$group->add(new Form_Select(
 		'dhgroup'.$key,
 		'*DH Group',
-		$p1enc['dhgroup'],
+		array_get_path($p1enc, 'dhgroup'),
 		$p1_dhgroups
 	))->setHelp($lastrow ? 'DH Group' : '')->setWidth(2);
 
@@ -988,7 +993,7 @@ foreach($pconfig['encryption']['item'] as $key => $p1enc) {
 	$group->add(new Form_Select(
 		'prfalgo'.$key,
 		'*PRF Algorithm',
-		$p1enc['prf-algorithm'],
+		array_get_path($p1enc, 'prf-algorithm'),
 		$p1_halgos
 	))->setHelp($lastrow ? 'PRF' : '')->setWidth(2);
 
diff --git a/src/etc/inc/ipsec.inc b/src/etc/inc/ipsec.inc
index 70cf909954..9229db94c1 100644
--- a/src/etc/inc/ipsec.inc
+++ b/src/etc/inc/ipsec.inc
@@ -2459,6 +2459,10 @@ function ipsec_setup_tunnels($gateways_status = false) {
 		if (isset($ph1ent['disabled'])) {
 			continue;
 		}
+		/* If the tunnel has no encryption algorithms, it isn't valid. */
+		if (count(array_get_path($ph1ent, 'encryption/item', [])) == 0) {
+			continue;
+		}
 		/* If the local source is invalid, skip this entry. */
 		$local_spec = ipsec_get_phase1_src($ph1ent, $gateways_status);
 		if (!$local_spec) {
@@ -2543,23 +2547,23 @@ function ipsec_setup_tunnels($gateways_status = false) {
 		}
 
 		/* Setup IKE proposals */
-		if (is_array($ph1ent['encryption']['item'])) {
-			$ciphers = array();
-			foreach($ph1ent['encryption']['item'] as $p1enc) {
-				if (!is_array($p1enc['encryption-algorithm']) ||
-						empty($p1enc['encryption-algorithm']['name']) ||
-						empty($p1enc['hash-algorithm'])) {
-					continue;
-				}
-				if ($ph1ent['prfselect_enable'] != 'yes') {
-					$p1enc['prf-algorithm'] = false;
-				}
-				$ciphers[] = ipsec_setup_proposal_algo($p1enc['encryption-algorithm']['name'],
-									$p1enc['encryption-algorithm']['keylen'],
-									$p1enc['hash-algorithm'],
-									$p1enc['prf-algorithm'],
-									$p1enc['dhgroup']);
+		$ciphers = array();
+		foreach(array_get_path($ph1ent, 'encryption/item', []) as $p1enc) {
+			if (empty($p1enc) ||
+			    empty(array_get_path($p1enc, 'encryption-algorithm/name', '')) ||
+			    empty(array_get_path($p1enc, 'hash-algorithm'))) {
+				continue;
 			}
+			if ($ph1ent['prfselect_enable'] != 'yes') {
+				$p1enc['prf-algorithm'] = false;
+			}
+			$ciphers[] = ipsec_setup_proposal_algo($p1enc['encryption-algorithm']['name'],
+								$p1enc['encryption-algorithm']['keylen'],
+								$p1enc['hash-algorithm'],
+								$p1enc['prf-algorithm'],
+								$p1enc['dhgroup']);
+		}
+		if (!empty($ciphers)) {
 			$conn['proposals'] = implode(",", $ciphers);
 		}
 
