diff --git a/src/usr/local/www/firewall_aliases_import.php b/src/usr/local/www/firewall_aliases_import.php
index 18e55eb45e..822ce6c18d 100644
--- a/src/usr/local/www/firewall_aliases_import.php
+++ b/src/usr/local/www/firewall_aliases_import.php
@@ -50,8 +50,7 @@ if (empty($tab)) {
 $pgtitle = array(gettext("Firewall"), gettext("Aliases"), gettext("Bulk import"));
 $pglinks = array("", "firewall_aliases.php?tab=" . $tab, "@self");
 
-init_config_arr(array('aliases', 'alias'));
-$a_aliases = &$config['aliases']['alias'];
+$a_aliases = config_get_path('aliases/alias', []);
 
 if ($_POST) {
 	$reqdfields = explode(" ", "name aliasimport");
@@ -77,7 +76,7 @@ if ($_POST) {
 	}
 
 	/* check for name interface description conflicts */
-	foreach ($config['interfaces'] as $interface) {
+	foreach (config_get_path('interfaces', []) as $interface) {
 		if ($interface['descr'] == $_POST['name']) {
 			$input_errors[] = gettext("An interface description with this name already exists.");
 			break;
@@ -85,11 +84,9 @@ if ($_POST) {
 	}
 
 	/* Is the description already used as an interface group name? */
-	if (is_array($config['ifgroups']['ifgroupentry'])) {
-		foreach ($config['ifgroups']['ifgroupentry'] as $ifgroupentry) {
-			if ($ifgroupentry['ifname'] == $_POST['name']) {
-				$input_errors[] = gettext("Sorry, an interface group with this name already exists.");
-			}
+	foreach (config_get_path('ifgroups/ifgroupentry', []) as $ifgroupentry) {
+		if ($ifgroupentry['ifname'] == $_POST['name']) {
+			$input_errors[] = gettext("An interface group with this name already exists.");
 		}
 	}
 
@@ -175,6 +172,7 @@ if ($_POST) {
 
 		// Sort list
 		$a_aliases = msort($a_aliases, "name");
+		config_set_path('aliases/alias', $a_aliases);
 
 		if (write_config(gettext("Imported a firewall alias."))) {
 			mark_subsystem_dirty('aliases');
