commit 345ede29dbfad4afdf6b1d7afd73213158b41cce
Author: jim-p <jimp@netgate.com>
Date:   Thu Nov 9 11:49:07 2023 -0500

    Fix PHP error when saving PPP interface w/o config. Issue #14949

diff --git a/src/usr/local/www/interfaces.php b/src/usr/local/www/interfaces.php
index e89bc61bac..b1cb05a640 100755
--- a/src/usr/local/www/interfaces.php
+++ b/src/usr/local/www/interfaces.php
@@ -1993,7 +1993,7 @@ $shortcut_section = "interfaces";
 $types4 = ["ppp" => gettext("PPP"), "pppoe" => gettext("PPPoE"), "pptp" => gettext("PPTP"), "l2tp" => gettext("L2TP")];
 
 if (!in_array(array_get_path($pconfig, 'type'), ["ppp", "pppoe", "pptp", "l2tp"]) ||
-   !array_key_exists(array_get_path($a_ppps, "{$pppid}/ports", []), get_configured_interface_list_by_realif())) {
+    !array_intersect_key(explode(",", array_get_path($a_ppps, "{$pppid}/ports", "")), get_configured_interface_list_by_realif())) {
 	$types4 = array_merge(["none" => gettext("None"), "staticv4" => gettext("Static IPv4"), "dhcp" => gettext("DHCP")], $types4);
 }
 
