commit 37884c3039756e3868b7f995a3329793b771a6b0
Author: jim-p <jimp@netgate.com>
Date:   Tue Feb 10 11:13:02 2026 -0500

    Correct trusted CA path. Fixes #16688

diff --git a/src/etc/inc/certs.inc b/src/etc/inc/certs.inc
index 1ebf1992c1..5299450e39 100644
--- a/src/etc/inc/certs.inc
+++ b/src/etc/inc/certs.inc
@@ -2552,8 +2552,8 @@ function cert_notify_expiring() {
  ******/
 
 function ca_setup_trust_store() {
-	/* This directory is trusted by OpenSSL on FreeBSD by default */
-	$trust_store_directory = '/usr/local/etc/ssl/certs';
+	/* Source for local CA certificates which will be trusted after "certctl rehash" runs */
+	$trust_store_directory = '/usr/local/share/certs/trusted';
 
 	/* Create the directory if it does not already exist, and clean it up if it does. */
 	safe_mkdir($trust_store_directory);
