diff --git a/src/usr/local/www/services_acb.php b/src/usr/local/www/services_acb.php
index 7c6f846130..69ae82c5f3 100644
--- a/src/usr/local/www/services_acb.php
+++ b/src/usr/local/www/services_acb.php
@@ -39,18 +39,18 @@ $exp_sep = '||';
 $decrypt_password = config_get_path('system/acb/encryption_password');
 
 // Defined username. Username must be sent lowercase. See Redmine #7127 and Netgate Redmine #163
-$username = strtolower($config['system']['acb']['gold_username']);
+$username = strtolower(config_get_path('system/acb/gold_username'));
 $password = config_get_path('system/acb/gold_password');
 
 // Set hostname
 if ($_REQUEST['hostname']) {
 	$hostname = $_REQUEST['hostname'];
 } else {
-	$hostname = $config['system']['hostname'] . "." . $config['system']['domain'];
+	$hostname = config_get_path('system/hostname') . "." . config_get_path('system/domain');
 }
 
 // Hostname of local machine
-$myhostname = $config['system']['hostname'] . "." . $config['system']['domain'];
+$myhostname = config_get_path('system/hostname') . "." . config_get_path('system/domain');
 
 if (!$decrypt_password) {
 	Header("Location: /services_acb_settings.php");
@@ -412,7 +412,7 @@ print('</div>');
 
 			<?php
 				$counter = 0;
-				if ($config['system']['acb']['reverse'] == "yes"){
+				if (config_get_path('system/acb/reverse') == "yes"){
 					$confvers = array_reverse($confvers);
 				}
 
diff --git a/src/usr/local/www/services_acb_backup.php b/src/usr/local/www/services_acb_backup.php
index f03d9b9779..4297a6b755 100644
--- a/src/usr/local/www/services_acb_backup.php
+++ b/src/usr/local/www/services_acb_backup.php
@@ -72,7 +72,7 @@ $tab_array[] = array("Restore", false, "/services_acb.php");
 $tab_array[] = array("Backup now", true, "/services_acb_backup.php");
 display_top_tabs($tab_array);
 
-$form = new Form("Backup", $config['system']['acb']['enable'] === "yes");
+$form = new Form("Backup", config_get_path('system/acb/enable') === "yes");
 
 $section = new Form_Section('Backup Details');
 
@@ -105,7 +105,7 @@ print($form);
 events.push(function() {
 	$(form).submit(function(e) {
 		e.preventDefault();
-		encpwd = '<?=$config['system']['acb']['encryption_password']?>';
+		encpwd = '<?=config_get_path("system/acb/encryption_password")?>';
 		if ( encpwd.length === 0) {
 			alert('<?=gettext("No encryption password has been set")?>');
 		} else if ($('#devkey').val().length === 0 ) {
