commit 587dcb6ae708d144023879b509a1da9dd5dd0723
Author: jim-p <jimp@netgate.com>
Date:   Wed Mar 6 14:49:28 2024 -0500

    Don't make OpenVPN routes for empty nets. Fixes #14919

diff --git a/src/etc/inc/openvpn.inc b/src/etc/inc/openvpn.inc
index 797f0f7b57..209d35de9f 100644
--- a/src/etc/inc/openvpn.inc
+++ b/src/etc/inc/openvpn.inc
@@ -2294,6 +2294,15 @@ function openvpn_gen_routes($value, $ipproto = "ipv4", $push = false, $iroute =
 		return "";
 	}
 	$tmpnetworks = explode(',', $value);
+
+	/* Remove empty entries from network array, which otherwise may lead to
+	 * invalid route statements.
+	 * https://redmine.pfsense.org/issues/14919
+	 */
+	$tmpnetworks = array_filter($tmpnetworks, function ($tnet) {
+		return !empty(trim($tnet));
+	});
+
 	foreach ($tmpnetworks as $network) {
 		if (is_alias($network)) {
 			foreach (alias_to_subnets_recursive($network, true) as $net) {
