commit 6a92af14584d22f077e1421e952674f880cd5b6c
Author: jim-p <jimp@netgate.com>
Date:   Tue Apr 1 15:22:07 2025 -0400

    Validation and output encoding of WOL interface. Fixes #16116

diff --git a/src/usr/local/www/services_wol.php b/src/usr/local/www/services_wol.php
index f3b8999872..44d9df7bc3 100644
--- a/src/usr/local/www/services_wol.php
+++ b/src/usr/local/www/services_wol.php
@@ -78,7 +78,7 @@ if ($_POST['Submit'] || $_POST['mac']) {
 		$input_errors[] = gettext("A valid MAC address must be specified.");
 	}
 
-	if (!$if) {
+	if (!$if || !array_key_exists($if, get_configured_interface_with_descr())) {
 		$input_errors[] = gettext("A valid interface must be specified.");
 	}
 
@@ -190,10 +190,10 @@ print $form;
 					<?php foreach (config_get_path('wol/wolentry', []) as $i => $wolent): ?>
 						<tr>
 							<td>
-								<?=convert_friendly_interface_to_friendly_descr($wolent['interface']);?>
+								<?=htmlspecialchars(convert_friendly_interface_to_friendly_descr($wolent['interface']));?>
 							</td>
 							<td>
-								<a href="?mac=<?=$wolent['mac'];?>&amp;if=<?=$wolent['interface'];?>" usepost><?=strtolower($wolent['mac']);?></a>
+								<a href="?mac=<?=$wolent['mac'];?>&amp;if=<?=urlencode($wolent['interface']);?>" usepost><?=strtolower($wolent['mac']);?></a>
 							</td>
 							<td>
 								<?=htmlspecialchars($wolent['descr']);?>
@@ -201,7 +201,7 @@ print $form;
 							<td>
 								<a class="fa-solid fa-pencil"	title="<?=gettext('Edit Device')?>"	href="services_wol_edit.php?id=<?=$i?>"></a>
 								<a class="fa-solid fa-trash-can"	title="<?=gettext('Delete Device')?>" href="services_wol.php?act=del&amp;id=<?=$i?>" usepost></a>
-								<a class="fa-solid fa-power-off" title="<?=gettext('Wake Device')?>" href="?mac=<?=$wolent['mac'];?>&amp;if=<?=$wolent['interface'];?>" usepost></a>
+								<a class="fa-solid fa-power-off" title="<?=gettext('Wake Device')?>" href="?mac=<?=$wolent['mac'];?>&amp;if=<?=urlencode($wolent['interface']);?>" usepost></a>
 							</td>
 						</tr>
 					<?php endforeach?>
diff --git a/src/usr/local/www/services_wol_edit.php b/src/usr/local/www/services_wol_edit.php
index e2d04af440..c4a3797c59 100644
--- a/src/usr/local/www/services_wol_edit.php
+++ b/src/usr/local/www/services_wol_edit.php
@@ -70,6 +70,10 @@ if ($_POST['save']) {
 
 	do_input_validation($_POST, $reqdfields, $reqdfieldsn, $input_errors);
 
+	if (!$_POST['interface'] || !array_key_exists($_POST['interface'], get_configured_interface_with_descr())) {
+		$input_errors[] = gettext("A valid interface must be specified.");
+	}
+
 	/* normalize MAC addresses - lowercase and convert Windows-ized hyphenated MACs to colon delimited */
 	$_POST['mac'] = trim(strtolower(str_replace("-", ":", $_POST['mac'])));
 
diff --git a/src/usr/local/www/widgets/widgets/wake_on_lan.widget.php b/src/usr/local/www/widgets/widgets/wake_on_lan.widget.php
index 78d26ba299..70c576d0b0 100644
--- a/src/usr/local/www/widgets/widgets/wake_on_lan.widget.php
+++ b/src/usr/local/www/widgets/widgets/wake_on_lan.widget.php
@@ -47,7 +47,7 @@ $wolcomputers = config_get_path('wol/wolentry', []);
 // Constructs a unique key that will identify a WoL entry in the filter list.
 if (!function_exists('get_wolent_key')) {
 	function get_wolent_key($wolent) {
-		return ($wolent['interface'] . "|" . $wolent['mac']);
+		return (htmlspecialchars($wolent['interface']) . "|" . $wolent['mac']);
 	}
 }
 
@@ -117,7 +117,7 @@ if (count($wolcomputers) > 0):
 				<?= $wolent['mac'] ?>
 			</td>
 			<td>
-				<?= convert_friendly_interface_to_friendly_descr($wolent['interface']) ?>
+				<?= htmlspecialchars(convert_friendly_interface_to_friendly_descr($wolent['interface'])) ?>
 			</td>
 			<td>
 		<?php if ($status == 'expires'): ?>
@@ -129,7 +129,7 @@ if (count($wolcomputers) > 0):
 		<?php endif; ?>
 			</td>
 			<td>
-				<a href="services_wol.php?mac=<?= $wolent['mac'] ?>&amp;if=<?= $wolent['interface']?>" usepost>
+				<a href="services_wol.php?mac=<?= $wolent['mac'] ?>&amp;if=<?= urlencode($wolent['interface']) ?>" usepost>
 				<i class="fa-solid fa-power-off" data-toggle="tooltip" title="<?= gettext("Wake up!") ?>"></i>
 				</a>
 			</td>
