commit 6f59a7f9fdfe3703667819fcbbd8b6f8cbec0d9f
Author: jim-p <jimp@netgate.com>
Date:   Thu Feb 1 11:12:51 2024 -0500

    ACB: Validate+encode frequency value. Fixes #15224

diff --git a/src/usr/local/www/services_acb_settings.php b/src/usr/local/www/services_acb_settings.php
index 4f5701d1af..beb283f0d2 100644
--- a/src/usr/local/www/services_acb_settings.php
+++ b/src/usr/local/www/services_acb_settings.php
@@ -65,6 +65,10 @@ if (isset($_POST['save'])) {
 		}
 	}
 
+	if (!in_array($_POST['frequency'], ['cron', 'every'])) {
+		$input_errors[] = gettext("Invalid frequency value.");
+	}
+
 	if ($_POST['frequency'] === 'cron') {
 		if (!preg_match('/^[0-9\*\/\-\,]+$/', $_POST['minute'] . $_POST['hour'] . $_POST['day'] . $_POST['month'] . $_POST['dow']))  {
 			$input_errors[] = gettext("Schedule values may only contain 0-9 - , / *");
@@ -229,7 +233,7 @@ print $form;
 			hideClass("cronsched", ($(this).val() != 'cron'));
 		});
 
-		hideClass("cronsched", ("<?=$pconfig['frequency']?>" != 'cron'));
+		hideClass("cronsched", ("<?=htmlspecialchars($pconfig['frequency'])?>" != 'cron'));
 	});
 //]]>
 </script>
