commit b891c3a33aff74f4ded6176a78b22ed84821036a
Author: jim-p <jimp@netgate.com>
Date:   Mon Apr 29 11:06:41 2024 -0400

    Add boot method to sysinfo widget. Implements #15422

diff --git a/src/usr/local/www/widgets/widgets/system_information.widget.php b/src/usr/local/www/widgets/widgets/system_information.widget.php
index 5cd0d07813..2d580526aa 100644
--- a/src/usr/local/www/widgets/widgets/system_information.widget.php
+++ b/src/usr/local/www/widgets/widgets/system_information.widget.php
@@ -218,9 +218,11 @@ $temp_use_f = (isset($user_settings['widgets']['thermal_sensors-0']) && !empty($
 		unset($biosvendor);
 		unset($biosversion);
 		unset($biosdate);
+		unset($bootmethod);
 		$_gb = exec('/bin/kenv -q smbios.bios.vendor 2>/dev/null', $biosvendor);
 		$_gb = exec('/bin/kenv -q smbios.bios.version 2>/dev/null', $biosversion);
 		$_gb = exec('/bin/kenv -q smbios.bios.reldate 2>/dev/null', $biosdate);
+		$bootmethod = get_single_sysctl("machdep.bootmethod");
 		/* Only display BIOS information if there is any to show. */
 		if (!empty($biosvendor[0]) || !empty($biosversion[0]) || !empty($biosdate[0])):
 ?>
@@ -236,6 +238,9 @@ $temp_use_f = (isset($user_settings['widgets']['thermal_sensors-0']) && !empty($
 			<?php if (!empty($biosdate[0])): ?>
 				<?=gettext("Release Date: ");?><strong><?= date("D M j Y ",strtotime($biosdate[0]));?></strong><br/>
 			<?php endif; ?>
+			<?php if (!empty($bootmethod)): ?>
+				<?=gettext("Boot Method: ");?><strong><?= htmlspecialchars($bootmethod) ?></strong><br/>
+			<?php endif; ?>
 			</td>
 		</tr>
 <?php
