commit dcdb461733044d274c742832097b13a312045f37
Author: jim-p <jimp@netgate.com>
Date:   Wed Nov 15 11:41:20 2023 -0500

    Fix IPsec log value handling. Fixes #14990

diff --git a/src/etc/inc/ipsec.inc b/src/etc/inc/ipsec.inc
index 68bfad2243..44f3d85edc 100644
--- a/src/etc/inc/ipsec.inc
+++ b/src/etc/inc/ipsec.inc
@@ -1190,7 +1190,7 @@ function ipsec_get_loglevels() {
 	$levels = array();
 
 	foreach (array_keys($ipsec_log_cats) as $cat) {
-		if (!empty(config_get_path('ipsec/logging/' . $cat))) {
+		if (strlen(config_get_path('ipsec/logging/' . $cat)) > 0) {
 			$levels[$cat] = config_get_path('ipsec/logging/' . $cat);
 		} elseif (in_array($cat, array('ike', 'chd', 'cfg'))) {
 			$levels[$cat] = "2";
