commit ddb57f79e26e97e2a22f701016fc70a7d1c09ce4
Author: jim-p <jimp@netgate.com>
Date:   Mon Sep 25 12:09:22 2023 -0400

    Update DDNS split host+domain list. Fixes #14783

diff --git a/src/etc/inc/globals.inc b/src/etc/inc/globals.inc
index e1b63dcd06..aaf02523aa 100644
--- a/src/etc/inc/globals.inc
+++ b/src/etc/inc/globals.inc
@@ -218,7 +218,21 @@ $factory_default_checkipservice = [
 ];
 
 global $dyndns_split_domain_types;
-$dyndns_split_domain_types = ['namecheap', 'cloudflare', 'cloudflare-v6', 'gratisdns', 'cloudns', 'godaddy', 'godaddy-v6', 'linode', 'linode-v6'];
+$dyndns_split_domain_types = [
+	'cloudflare', 'cloudflare-v6',
+	'cloudns',
+	"digitalocean", "digitalocean-v6",
+	"gandi-livedns", "gandi-livedns-v6",
+	'godaddy', 'godaddy-v6',
+	'gratisdns',
+	'linode', 'linode-v6',
+	"mythicbeasts", "mythicbeasts-v6",
+	'namecheap',
+	"name.com", "name.com-v6",
+	"onecom", "onecom-v6",
+	'porkbun', 'porkbun-v6',
+	"yandex", "yandex-v6",
+];
 
 // pf tokens from FreeBSD source /sbin/pfctl/parse.y
 global $pf_reserved_keywords;
