commit f106b62cfbed279e8140ffa1edf535defb0221ab
Author: jim-p <jimp@netgate.com>
Date:   Thu Sep 14 10:26:47 2023 -0400

    Fix format of OpenVPN cached interface. Fixes #14781

diff --git a/src/etc/inc/openvpn.inc b/src/etc/inc/openvpn.inc
index 1e63e3dfbf..480ce033fa 100644
--- a/src/etc/inc/openvpn.inc
+++ b/src/etc/inc/openvpn.inc
@@ -1547,7 +1547,7 @@ function openvpn_reconfigure($mode, $settings) {
 	file_put_contents($fpath, $conf);
 	unset($conf);
 	$fpath = "{$g['openvpn_base']}/{$mode_id}/interface";
-	file_put_contents($fpath, $interface);
+	file_put_contents($fpath, get_failover_interface($interface));
 	$fpath = "{$g['openvpn_base']}/{$mode_id}/connuserlimit";
 	file_put_contents($fpath, $connlimit);
 	//chown($fpath, 'nobody');
