commit f79dfc8c6b8d51a7781f9fe886eb69e5bd9dde62
Author: Marcos Mendoza <mmendoza@netgate.com>
Date:   Thu Dec 12 17:11:58 2024 -0600

    Check subqueues when determining pipe slot limit. Fix #13662

diff --git a/src/etc/inc/shaper.inc b/src/etc/inc/shaper.inc
index 569fa3acaf..8d5f109b63 100644
--- a/src/etc/inc/shaper.inc
+++ b/src/etc/inc/shaper.inc
@@ -5454,6 +5454,14 @@ function filter_generate_dummynet_rules() {
 		if ($this_qlimit > $max_qlimit) {
 			$max_qlimit = $this_qlimit;
 		}
+		if (is_array($dn->subqueues)) {
+			foreach ($dn->subqueues as $queue) {
+				$this_qlimit = $queue->GetQlimit();
+				if ($this_qlimit > $max_qlimit) {
+					$max_qlimit = $this_qlimit;
+				}
+			}
+		}
 	}
 	if (!is_numericint($max_qlimit)) {
 		$max_qlimit = "100";
