commit faf9f096448c3d18ba291901e391270036ab47c7
Author: jim-p <jimp@netgate.com>
Date:   Mon Mar 25 13:43:16 2024 -0400

    Fix variable typo. Fixes #14996

diff --git a/src/etc/inc/services.inc b/src/etc/inc/services.inc
index 58af033d57..cd1c01a665 100644
--- a/src/etc/inc/services.inc
+++ b/src/etc/inc/services.inc
@@ -1408,7 +1408,7 @@ function services_kea4_configure()
 			if (is_array($poolconf['winsserver']) && $poolconf['winsserver'][0] && ($poolconf['winsserver'][0] != $dhcpifconf['winsserver'][0])) {
 				$keapool['option-data'][] = [
 					'name' => 'netbios-name-servers',
-					'data' => implode(', ', $poolconf['winserver'])
+					'data' => implode(', ', $poolconf['winsserver'])
 				];
 				$keapool['option-data'][] = [
 					'name' => 'netbios-node-type',
