diff --git a/src/etc/sshd b/src/etc/sshd
index 90380a8d62..1ab5c6e900 100755
--- a/src/etc/sshd
+++ b/src/etc/sshd
@@ -115,8 +115,8 @@ if (isset($config['system']['ssh']['sshdagentforwarding'])) {
 	$sshconf .= "AllowAgentForwarding no\n";
 }
 $sshconf .= "X11Forwarding no\n";
-$sshconf .= "Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr\n";
-$sshconf .= "MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,umac-128@openssh.com\n";
+$sshconf .= "Ciphers aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr\n";
+$sshconf .= "MACs hmac-sha2-512,hmac-sha2-256,umac-128@openssh.com\n";
 $sshconf .= "# override default of no subsystems\n";
 $sshconf .= "Subsystem\tsftp\t/usr/libexec/sftp-server\n";
 
