diff -urN snort-2.9.9.0.orig/src/output-plugins/spo_csv.c snort-2.9.9.0/src/output-plugins/spo_csv.c
--- snort-2.9.9.0.orig/src/output-plugins/spo_csv.c	2016-06-07 03:47:49.000000000 -0400
+++ ./src/output-plugins/spo_csv.c	2017-08-17 19:55:33.000000000 -0400
@@ -69,7 +69,7 @@
 #include "sfutil/sf_textlog.h"
 #include "log_text.h"
 
-#define DEFAULT_CSV "timestamp,sig_generator,sig_id,sig_rev,msg,proto,src,srcport,dst,dstport,ethsrc,ethdst,ethlen,tcpflags,tcpseq,tcpack,tcpln,tcpwindow,ttl,tos,id,dgmlen,iplen,icmptype,icmpcode,icmpid,icmpseq"
+#define DEFAULT_CSV "timestamp,sig_generator,sig_id,sig_rev,msg,proto,src,srcport,dst,dstport,ethsrc,ethdst,ethlen,tcpflags,tcpseq,tcpack,tcpln,tcpwindow,ttl,tos,id,dgmlen,iplen,icmptype,icmpcode,icmpid,icmpseq,classification,priority"
 
 #define DEFAULT_FILE  "alert.csv"
 #define DEFAULT_LIMIT (128*M_BYTES)
@@ -90,6 +90,8 @@
     AlertCSVConfig *config;
 } AlertCSVData;
 
+/* Global variable used by Log output and other functions */
+extern OptTreeNode *otn_tmp;
 
 /* list of function prototypes for this preprocessor */
 static void AlertCSVInit(struct _SnortConfig *, char *);
@@ -499,6 +501,18 @@
                 TextLog_Print(log, "%s", tcpFlags);
             }
         }
+        else if (!strcasecmp("classification",type))
+        {
+		if (otn_tmp != NULL) {
+			if ((otn_tmp->sigInfo.classType != NULL) && (otn_tmp->sigInfo.classType->name != NULL))
+				TextLog_Print(log, "%s", otn_tmp->sigInfo.classType->name);
+			}
+        }
+        else if (!strcasecmp("priority",type))
+        {
+			if (otn_tmp != NULL)
+				TextLog_Print(log, "%d", otn_tmp->sigInfo.priority);
+		}
 
         if (num < numargs - 1)
             TextLog_Putc(log, ',');
